cassiopeia/simulator/BirthDeathFitnessSimulator.py

```diff
@@ -0,0 +1,399 @@
+        initial_birth_scale = 0.5
+        mutation_distribution = lambda: 1 if np.random.uniform() > 0.5 else 0
+        fitness_distribution = lambda: np.random.uniform(-1,1)
+        fitness_base = 2
+
+        bd_sim = BirthDeathFitnessSimulator(
+            birth_waiting_distribution,
+            initial_birth_scale,
+            death_waiting_distribution=death_waiting_distribution,
+            mutation_distribution=mutation_distribution,
+            fitness_distribution=fitness_distribution,
+            fitness_base=fitness_base,
+            num_extant=8
+        )
+        tree = bd_sim.simulate_tree()
+
+    Args:
+        birth_waiting_distribution: A function that samples waiting times
+            from the birth distribution. Determines how often births occur.
+            Must take a scale parameter as the input
+        initial_birth_scale: The initial scale parameter that is used at the
+            start of the experiment
+        death_waiting_distribution: A function that samples waiting times
+            from the death distribution. Determines how often deaths occur
+        mutation_distribution: A function that samples the number of
+            mutations that occur at a division event. If None, then no
+            mutations are sampled
+        fitness_distribution: One of the two elements in determining the
+            multiplicative coefficient of a fitness mutation. A function that
+            samples the exponential that the mutation base is raised by.
+            Determines the distribution of fitness mutation strengths. Must not
+            be None if mutation_distribution provided
+        fitness_base: One of the two elements in determining the
+            multiplicative strength of a fitness mutation. The base that is
+            raised by the value given by the fitness distribution. Determines
+            the base strength of fitness mutations. By default is e, Euler's
+            Constant
+        num_extant: Specifies the number of extant lineages existing at the
+            same time as a stopping condition for the experiment
+        experiment_time: Specifies the total time that the experiment runs as a
+            stopping condition for the experiment
+        collapse_unifurcations: Specifies whether to collapse unifurcations in
+            the tree resulting from pruning dead lineages
+    """
+
+    def __init__(
+        self,
+        birth_waiting_distribution: Callable[[float], float],
+        initial_birth_scale: float,
+        death_waiting_distribution: Optional[
+            Callable[[], float]
+        ] = lambda: np.inf,
+        mutation_distribution: Optional[Callable[[], int]] = None,
+        fitness_distribution: Optional[Callable[[], float]] = None,
+        fitness_base: float = np.e,
+        num_extant: Optional[int] = None,
+        experiment_time: Optional[float] = None,
+        collapse_unifurcations: bool = True,
+    ):
```

Comment:
I propose we add the option to add a `random_state` or `random_seed` (preferably a state) for determinism. This will be nice when distributing our code / simulations as well as for testing purposes below.

Reply:
The randomness is abstracted out into `birth_waiting_distribution, death_waiting_distribution, mutation_distribution, fitness_distribution` which are provided by the user, I don't think we need to deal with randomness at all in our code?

Reply:
Hi Matt, 
Currently I'm using random seeds. I think utilizing a random_state might be hard, because it has a set of built-in distributions that it samples from, but we want the BirthDeathFitnessSimulator to be able to use any given distribution. But you're right let me allow the user to pass in a seed as an input

Reply:
also Sebastian in regards to your comment, I think we can make deterministic distributions to pass in, but let's say we're using a function that defines the exponential with some scale. We'd still encounter randomness in the waiting times we draw from the exponential distribution.

Reply:
I'm not sure I understand, here's how I would create my `birth_waiting_distribution` as a user:
```
random_state = np.random.RandomState(1)
def birth_waiting_distribution():
    random_state.exponential()
```
BirthDeathFitnessSimulator does not need to handle randomness in this case, it will be deterministic. It's up to the user to make `birth_waiting_distribution`, etc. deterministic.

Reply:
Ah I see Sebastian. I think that it may take some of the burden off the user if we allow for a seed though, that way they don't have to create a randomstate for each function. And it also means that they don't need to know about the use/existence of randomstates

Reply:
To wrap up this conversation from earlier -- I agree with Sebastian that we should rely on users to pass in random states. However, I believe we all agreed that it would be nice to support the use of random seeds for reproducibility.

Reply:
Agreed.